### PR TITLE
Autocomplete history terms feature

### PR DESCRIPTION
**Pull request by chrisFrazier77**

### CODE REVIEW

File: packages/snap-controller/src/Autocomplete/AutocompleteController.ts [hunk 1]
```diff
@@ -338,8 +338,13 @@ export class AutocompleteController extends AbstractController {
 			},
 			keyUp: (e: KeyboardEvent): void => {
 				// ignore enter and escape keys
-				if (e?.keyCode == KEY_ENTER || e?.keyCode == KEY_ESCAPE) return;
-
+				if (e?.keyCode == KEY_ENTER || e?.keyCode == KEY_ESCAPE) {
+					if (e.keyCode == KEY_ENTER) {
+						const input = e.target as HTMLInputElement;
```
Comment (dkonieczek) [inline packages/snap-controller/src/Autocomplete/AutocompleteController.ts:343]: Maybe a question for @korgon 
Do you think the term that we save to history should be what was typed in the input or the suggested term? 
Or maybe only use the user typed in term if integrated spell correction is enabled?

File: packages/snap-preact-components/src/components/Organisms/Autocomplete/Autocomplete.tsx [hunk 2]
```diff
@@ -170,6 +170,7 @@ export const Autocomplete = observer((properties: AutocompleteProps): JSX.Elemen
 		// default props
 		termsTitle: '',
 		trendingTitle: 'Popular Searches',
+		historyTitle: 'Previously Searched',
```
Comment (dkonieczek) [inline packages/snap-preact-components/src/components/Organisms/Autocomplete/Autocomplete.tsx:173]: `hideHistory` has a default value of `true` in the stories file however it looks to have a default value of `false` in the component file as it's not defined here.

File: packages/snap-preact-components/src/components/Organisms/Autocomplete/Autocomplete.tsx [hunk 3]
```diff
@@ -454,7 +470,33 @@ export const Autocomplete = observer((properties: AutocompleteProps): JSX.Elemen
 												</div>
 											))}
 										</div>
-									</>
+									</div>
+								) : null}
+
+								{history.length && !hideHistory ? (
+									<div className="ss__autocomplete__terms__history">
```
Comment (dkonieczek) [inline packages/snap-preact-components/src/components/Organisms/Autocomplete/Autocomplete.tsx:477]: May want to not render the title divs if someone provides `<Autocomplete historyTitle={''}>`
Terms and Trending has this:
```
{trendingTitle ? (
    <div className="ss__autocomplete__title ss__autocomplete__title--trending">
        <h5>{trendingTitle}</h5>
    </div>
) : null}
```

File: packages/snap-preact-components/src/components/Organisms/Autocomplete/Autocomplete.tsx [hunk 4]
```diff
@@ -454,7 +470,33 @@ export const Autocomplete = observer((properties: AutocompleteProps): JSX.Elemen
...
+										<div className="ss__autocomplete__title ss__autocomplete__title--history">
+											<h5>{historyTitle}</h5>
+										</div>
+										<div className="ss__autocomplete__terms__options ss__autocomplete__history__options">
+											{history.map((term: any) => (
```
Comment (dkonieczek) [inline packages/snap-preact-components/src/components/Organisms/Autocomplete/Autocomplete.tsx:482]: Can we remove the `any` type here to match `trending` and `terms` if it's not possible to use the `Term` type in all of these.

**Review by dkonieczek — CHANGES_REQUESTED** [review]

File: packages/snap-controller/src/Autocomplete/AutocompleteController.ts (hunk 1 above)
Comment (korgon) [reply packages/snap-controller/src/Autocomplete/AutocompleteController.ts:343]: Yea, we should use the term (query) that we send with our redirect to the search page - in any case when using our autocomplete the first suggested term will contain the query that we direct to the search page with.

File: packages/snap-preact-components/src/components/Organisms/Autocomplete/Autocomplete.tsx (hunk 2 above)
Comment (chrisFrazier77) [reply packages/snap-preact-components/src/components/Organisms/Autocomplete/Autocomplete.tsx:173]: good catch, it was true by default initially, but refactored to undefined.

**Review by korgon — CHANGES_REQUESTED** [review]
It looks like we are only saving to history when the enter key is pressed - we should consider saving in all circumstances:
* when enter key is pressed
* when form is submitted
* when terms are clicked

Couple of thoughts:
Maybe instead of using the autocomplete controller/store to save history we should save them inside of the search controller? For example - when a search is run in a search controller we would save the query to history (perhaps on the SearchController search method execution). This would ensure that the history is accurate (would even build history via URLs). This could be done on a per siteId basis - and the AutocompleteController would simply instantiate a HistoryStore passing it a siteId in the config.

I'd like to see the history store be self contained and managed - by that I mean it should contain the methods to add to history, retrieve history or reset it. In instantiation it can have configuration for the limit and perhaps the storage name. This would make the history store more stand alone similar to other stores (pretty much just a wrapper around the storage store with special methods).

**Review by korgon — APPROVED** [review]

**Review by dkonieczek — APPROVED** [review]

**Pull request merged**